### PR TITLE
Fix masking error when using mlp_bias=True

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

Bug fix for b/497864549: [XL ML] NaN training loss for GPT-OSS SFT.

GPT-OSS is currently experiencing NaN training loss after one step when using `expert_parallelism > 1`. The bug was caused by lack of masking in the mlp bias inside of expert computation. Since no other models use the mlp bias they didn't experience this issue. 

## Issue
Currently, when `expert_parallelism > 1` we introduce a buffer to store the output of `ragged_all_to_all`, which may contain padding along the token axis. If padding values are not masked after adding the MLP bias, JAX will include them in the gradient computation. `jnp.where` allows us to disconnect the padding values from the backwards graph during bias gradient calculation.

# Tests

1. Ran `pre_train.train` on `GPT_OSS` with `expert_parallelism > 1`: https://paste.googleplex.com/6279941382602752
2. logits matching against golden logits: https://paste.googleplex.com/6165440129794048

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by RissyRan — COMMENTED** [review]
Thanks! Could you have a logits correctness check with EP>1 using GPT-OSS?

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -1304,6 +1301,7 @@ def get_active_sharding_axes(pspec_dim_axes, tensor_dim_index):
         layer_w0 = jax.lax.psum(layer_w0, "tensor_transpose")
       if self.config.mlp_bias:
         layer_w0 = layer_w0 + w0_bias
+        layer_w0 = jnp.where(mask[:, None], layer_w0, 0)
```
Comment (RissyRan) [inline src/maxtext/layers/moe.py:1304]: Is default mask defined without EP sharding?

Comment (snehalv2002) [reply src/maxtext/layers/moe.py]: Yes the mask is defined even with no EP. [`mask`](https://github.com/AI-Hypercomputer/maxtext/pull/3699/changes#diff-e080eae53990c3bcfd6cd1f60db63f57d80988dfe8715b4ffbc09fbcabb5d8f3R1240) is defined outside the EP > 1 block

**Review by RissyRan — APPROVED** [review]
Thanks for the fix!

File: src/maxtext/layers/moe.py [hunk 2]
```diff
@@ -2138,12 +2146,13 @@ def _moe_body(
         )
       else:
         x, routing, route_metadata = route(x, logits, pre_bias_logits, rngs, input_ids=sharded_input_ids)
+        mask = jnp.arange(x.shape[0]) < jnp.sum(routing.group_sizes)
```
Comment (gagika) [inline src/maxtext/layers/moe.py]: One thing before merge:

`mask` assumes valid rows = `sum(routing.group_sizes)`, which only holds when ring-of-experts is off. With ring, `get_gmm_for_local_experts` uses `routing.local_group_sizes` (L1942) or a `group_offset`-ed slice (L1950), while `routing.group_sizes` stays the **global** bincount — so `sum(...) >= x.shape[0]`, the mask is all-True, and `use_ring_of_experts=True` + `mlp_bias=True` still NaNs. It doesn't fail loudly, it just quietly does nothing.

Suggestion: compute the valid-row count next to `get_gmm_for_local_experts`, off the same branches, so the two can't drift apart:

```python
def valid_token_count(x, routing, route_metadata):
  """Rows of `x` the gmm actually computes; the rest is ragged-buffer padding."""
  num_ep = self.get_expert_parallelism_size()
  num_experts_per_shard = self.config.num_experts // num_ep
  if self.config.use_ring_of_experts and x.shape[0] < routing.sorted_selected_experts.shape[0]:
    return jnp.sum(routing.local_group_sizes)
  if self.config.use_ragged_sort and self.config.use_ring_of_experts:
    experts_start = route_metadata.expert_shard_id * num_experts_per_shard
    return jnp.sum(
        jax.lax.dynamic_slice_in_dim(routing.group_sizes, experts_start, num_experts_per_shard)
    )
  return jnp.sum(routing.group_sizes)
```

Comment (snehalv2002) [reply src/maxtext/layers/moe.py]: Done

**Review by gagika — APPROVED** [review]
